### PR TITLE
Cns 786 spec readme

### PR DESCRIPTION
**Pull request by Yaroms**

### CODE REVIEW

File: x/spec/README.md [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+* [Parameters](#parameters)
+* [Queries](#queries)
+* [Transactions](#transactions)
+* [Proposals](#proposals)
+
+## Concepts
+
+### Spec
+
+A chain spec consists of general properties of the chain and a list of interface it supports. To make specs simple to create and maintain specs can import api's from another, for example, X testnet spec can Import from X mainnet spec and thus not need to redifine all of the interfaces.
```
Comment (oren-lava) [inline x/spec/README.md:28]: AI:

A chain spec consists of general properties of the chain and a list of interfaces it supports. To simplify the creation and maintenance of specs, they can import APIs from another spec. For example, the X testnet spec can import from the X mainnet spec, eliminating the need to redefine all of the interfaces.

File: x/spec/README.md [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    AllowedBlockLagForQosSync     int64                 // defines the accepted blocks a provider can be behind the chain without QOS degradation
+	BlockLastUpdated              uint64                // the last block this spec was updated on chain
+    ReliabilityThreshold          uint32                // this determines the amount of data reliability for the chain
+	DataReliabilityEnabled        bool                  // enables/disables data reliability for the chain
+	BlockDistanceForFinalizedData uint32                              
+	BlocksInFinalizationProof     uint32                              
+}
+```
+Note, the `Coin` type is from Cosmos-SDK (`cosmos.base.v1beta1.Coin`).
+Note, the `Dec` type is from Cosmos-SDK math LegacyDec.
```
Comment (oren-lava) [inline x/spec/README.md:52]: from Cosmos-SDK math module `LegacyDec` (added backticks to look like code)

File: x/spec/README.md [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+
+## Concepts
+
+### Spec
+
+A chain spec consists of general properties of the chain and a list of interface it supports. To make specs simple to create and maintain specs can import api's from another, for example, X testnet spec can Import from X mainnet spec and thus not need to redifine all of the interfaces.
+
+```
+type Spec struct {
+	Index                         string                // chain index (index key for the chain)
```
Comment (oren-lava) [inline x/spec/README.md:32]: chain index (index key for the chain) -> chain unique index

File: x/spec/README.md (hunk 2 above)
Comment (oren-lava) [inline x/spec/README.md:52]: put extra line between the "Note" sentences to make it appear as a new line. Also, in the second "Note" I would replace it with: "Also" or "Also note"

File: x/spec/README.md [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+	AverageBlockTime              int64                 // average block time of the chain in msec
+	MinStakeProvider              Coin                  // min stake for a provider to be active in the chain
+	ProvidersTypes                Spec_ProvidersTypes   // determines if the spec is for lava or others 
+	Imports                       []string              // list of chains to import ApiCollections from
+	ApiCollections                []*ApiCollection      // list of ApiCollections that defines all the interfaces and api's of the chain
+	Contributor                   []string              // list of contributers (public lava address {lava@...})
+	ContributorPercentage         *Dec                  // the percentage of coins the contributers will get from each reward a provider get
+	Shares                        uint64                // factor for bonus rewards at the end of the month (see rewards module)
+    AllowedBlockLagForQosSync     int64                 // defines the accepted blocks a provider can be behind the chain without QOS degradation
```
Comment (oren-lava) [inline x/spec/README.md:43]: align to other fields (see README preview)

File: x/spec/README.md [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+	BlockLastUpdated              uint64                // the last block this spec was updated on chain
+    ReliabilityThreshold          uint32                // this determines the amount of data reliability for the chain
+	DataReliabilityEnabled        bool                  // enables/disables data reliability for the chain
```
Comment (oren-lava) [inline x/spec/README.md:46]: align to other fields (see README preview)

File: x/spec/README.md [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+
+### Spec
+
+A chain spec consists of general properties of the chain and a list of interface it supports. To make specs simple to create and maintain specs can import api's from another, for example, X testnet spec can Import from X mainnet spec and thus not need to redifine all of the interfaces.
+
+```
+type Spec struct {
+	Index                         string                // chain index (index key for the chain)                             
+	Name                          string                // description string of the spec
+	Enabled                       bool                  // spec enabled/disable (base specs are usually disabled, like cosmos base specs)
```
Comment (oren-lava) [inline x/spec/README.md:34]: parentheses in the comment is irrelevant

File: x/spec/README.md [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+	AverageBlockTime              int64                 // average block time of the chain in msec
+	MinStakeProvider              Coin                  // min stake for a provider to be active in the chain
+	ProvidersTypes                Spec_ProvidersTypes   // determines if the spec is for lava or others 
+	Imports                       []string              // list of chains to import ApiCollections from
+	ApiCollections                []*ApiCollection      // list of ApiCollections that defines all the interfaces and api's of the chain
```
Comment (oren-lava) [inline x/spec/README.md:39]: api's -> APIs

File: x/spec/README.md [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+A chain spec consists of general properties of the chain and a list of interface it supports. To make specs simple to create and maintain specs can import api's from another, for example, X testnet spec can Import from X mainnet spec and thus not need to redifine all of the interfaces.
+
+```
+type Spec struct {
+	Index                         string                // chain index (index key for the chain)                             
+	Name                          string                // description string of the spec
+	Enabled                       bool                  // spec enabled/disable (base specs are usually disabled, like cosmos base specs)
+	AverageBlockTime              int64                 // average block time of the chain in msec
+	MinStakeProvider              Coin                  // min stake for a provider to be active in the chain
+	ProvidersTypes                Spec_ProvidersTypes   // determines if the spec is for lava or others
```
Comment (oren-lava) [inline x/spec/README.md:37]: others? what others?
can you elaborate below the struct?

File: x/spec/README.md [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+type Spec struct {
+	Index                         string                // chain index (index key for the chain)                             
+	Name                          string                // description string of the spec
+	Enabled                       bool                  // spec enabled/disable (base specs are usually disabled, like cosmos base specs)
+	AverageBlockTime              int64                 // average block time of the chain in msec
+	MinStakeProvider              Coin                  // min stake for a provider to be active in the chain
+	ProvidersTypes                Spec_ProvidersTypes   // determines if the spec is for lava or others 
+	Imports                       []string              // list of chains to import ApiCollections from
+	ApiCollections                []*ApiCollection      // list of ApiCollections that defines all the interfaces and api's of the chain
+	Contributor                   []string              // list of contributers (public lava address {lava@...})
```
Comment (oren-lava) [inline x/spec/README.md:40]: you did not define anywhere what a contributor is

File: x/spec/README.md [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+	ContributorPercentage         *Dec                  // the percentage of coins the contributers will get from each reward a provider get
+	Shares                        uint64                // factor for bonus rewards at the end of the month (see rewards module)
+    AllowedBlockLagForQosSync     int64                 // defines the accepted blocks a provider can be behind the chain without QOS degradation
+	BlockLastUpdated              uint64                // the last block this spec was updated on chain
+    ReliabilityThreshold          uint32                // this determines the amount of data reliability for the chain
+	DataReliabilityEnabled        bool                  // enables/disables data reliability for the chain
+	BlockDistanceForFinalizedData uint32
```
Comment (oren-lava) [inline x/spec/README.md:47]: no explanation for finalization fields

File: x/spec/README.md [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+* [Proposals](#proposals)
+
+## Concepts
+
+### Spec
+
+A chain spec consists of general properties of the chain and a list of interface it supports. To make specs simple to create and maintain specs can import api's from another, for example, X testnet spec can Import from X mainnet spec and thus not need to redifine all of the interfaces.
+
+```
+type Spec struct {
```
Comment (oren-lava) [inline x/spec/README.md:31]: The usual format is "struct X" and not "type X struct". I thought it's more readable

File: x/spec/README.md [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+
+rules:
+* an import cycle of specs is prohibited
+* specs can override/disable/add api's from the imported api collection
+* specs imports the verificaions, they can also be overridden and sometimes it is a must (for example chain-id value must be overwritten)
+
+
+## Parameters
+
+The plans module does not contain parameters.
```
Comment (oren-lava) [inline x/spec/README.md:178]: plans -> spec

File: x/spec/README.md [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+
+## Queries
+
+The plans module supports the following queries:
```
Comment (oren-lava) [inline x/spec/README.md:182]: plans -> spec

File: x/spec/README.md [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+| ----------        | ---------------   | ----------------------------------------------|
+| `list-spec`       | none              | shows all the specs                           |
+| `params`          | none              | show the params of the module                 |
+| `show-all-chains` | none              | shows all the specs with minimal info         |
+| `show-chain-info` | chainid           | shows a spec with minimal info                |
+| `show-spec`       | chainid           | shows a full spec                             |
+
+## Transactions
+
+The plans module does not support any transactions.
```
Comment (oren-lava) [inline x/spec/README.md:194]: plans -> spec

File: x/spec/README.md [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+
+## Proposals
+
+The plans module provides a proposal to add/overwrite a spec to the chain
```
Comment (oren-lava) [inline x/spec/README.md:198]: plans -> spec

File: x/spec/README.md [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+	MinStakeProvider              Coin                  // min stake for a provider to be active in the chain
+	ProvidersTypes                Spec_ProvidersTypes   // determines if the spec is for lava or others 
+	Imports                       []string              // list of chains to import ApiCollections from
+	ApiCollections                []*ApiCollection      // list of ApiCollections that defines all the interfaces and api's of the chain
+	Contributor                   []string              // list of contributers (public lava address {lava@...})
+	ContributorPercentage         *Dec                  // the percentage of coins the contributers will get from each reward a provider get
+	Shares                        uint64                // factor for bonus rewards at the end of the month (see rewards module)
+    AllowedBlockLagForQosSync     int64                 // defines the accepted blocks a provider can be behind the chain without QOS degradation
+	BlockLastUpdated              uint64                // the last block this spec was updated on chain
+    ReliabilityThreshold          uint32                // this determines the amount of data reliability for the chain
```
Comment (oren-lava) [inline x/spec/README.md:45]: what is "amount of data reliability"?

**Review by oren-lava — CHANGES_REQUESTED** [review]

File: x/spec/README.md [hunk 17]
```diff
@@ -0,0 +1,298 @@
...
+
+### Events
+
+
+The plans module has the following events:
+| Event             | When it happens       |
+| ----------        | --------------- |
+| `spec_add`        | a successful addition of a spec  |
+| `spec_modify`     | a successful modification of an existing spec   |
+| `spec_refresh`    | a spec was redreshed since it had a imported spec modified|
```
Comment (oren-lava) [inline x/spec/README.md:298]: redreshed -> refreshed

**Review by oren-lava — CHANGES_REQUESTED** [review]

File: x/spec/README.md (hunk 17 above)
Comment (Yaroms) [reply x/spec/README.md:298]: done

File: x/spec/README.md [hunk 18]
```diff
@@ -0,0 +1,298 @@
...
+	InheritanceApis []*CollectionData   // list of other ApiCollection to inherite from
+	ParseDirectives []*[ParseDirective](#parsedirective)   // list of parsing instructions of specific api's
+	Extensions      []*Extension        // list of extensions that providers can support in addition to the basic behaviour (for example, archive node)
+	Verifications   []*Verification     // list of verifications that providers must pass to make sure they provide full functionality
+}
+```
+
+### CollectionData
+
+CollectionData defines the api properties and acts as a unique key for the [api collection](#apicollection)
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/spec/README.md:77]: done

File: x/spec/README.md [hunk 19]
```diff
@@ -0,0 +1,298 @@
...
+}
+```
+
+The `ApiInterface` field defines the API interface on which the limitations are applied. The available API interfaces for a chain are defined in the chain's spec. Overall, the API interfaces can be: `jsonrpc`, `rest`, `tendermintrpc` and `grpc`.
+
+The `InternalPath` field is utilized for chains that have varying RPC API sets in different internal paths. Avalanche is a prime example of such a chain, consisting of three distinct subchains (or subnets) designed for different applications. For instance, Avalanche's C-Chain is dedicated to smart contracts, while Avalanche's X-Chain facilitates the sending and receiving of funds. For further information on how to define this field, please consult the Avalanche (AVAX) specification.
+
+The `Type` field lets the user define APIs that have different functionalities depending on their type. the valid types are: `GET` and `POST`. An example of such API is Cosmos' `/cosmos/tx/v1beta1/txs` API. If it's sent as a `GET` request, it fetches transactions by event and if it's sent as a `POST` request, it sends a transaction.
+
+The `AddOn` field lets you use additional optional APIs like debug, trace,
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/spec/README.md:94]: done

File: x/spec/README.md [hunk 20]
```diff
@@ -0,0 +1,298 @@
...
+            "subscription": false,
+            "stateful": 0
+        },
+        "extra_compute_units": 0
+    },
+```
+
+### ParseDirective
+
+ParseDirective is a struct that defines a function needed by the provider in a generic way. it describes how for a specific api collection how to get information from the node. for example, how to get the latest block of an EVM node.
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/spec/README.md:136]: will do

File: x/spec/README.md [hunk 21]
```diff
@@ -0,0 +1,298 @@
...
+ApiCollection is a struct that defines an interface, such as REST, JSON, etc., along with all of its APIs and properties.
+
+```
+type ApiCollection struct {
+	Enabled         bool                // enables/disables the collection
+	CollectionData  CollectionData      // defines the properties of the collection, also acts as a unique key
+	Apis            []*Api              // list of api's in the collection
+	Headers         []*Header           // list of headers supported by the interface and their behaviour
+	InheritanceApis []*CollectionData   // list of other ApiCollection to inherite from
+	ParseDirectives []*[ParseDirective](#parsedirective)   // list of parsing instructions of specific api's
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/spec/README.md:69]: done

File: x/spec/README.md [hunk 22]
```diff
@@ -0,0 +1,298 @@
...
+### ApiCollection
+
+ApiCollection is a struct that defines an interface, such as REST, JSON, etc., along with all of its APIs and properties.
+
+```
+type ApiCollection struct {
+	Enabled         bool                // enables/disables the collection
+	CollectionData  CollectionData      // defines the properties of the collection, also acts as a unique key
+	Apis            []*Api              // list of api's in the collection
+	Headers         []*Header           // list of headers supported by the interface and their behaviour
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/spec/README.md:67]: done

File: x/spec/README.md [hunk 23]
```diff
@@ -0,0 +1,298 @@
...
+Api define a specific api in the api collection
+
+```
+type Api struct {
+	Enabled           bool          // enable/disable the api
+	Name              string        // api name
+	ComputeUnits      uint64        // the amount of cu of this api (can be defined as the "price" of using this api)
+	ExtraComputeUnits uint64        // not used
+	Category          SpecCategory  // defines the property of the api
+	BlockParsing      BlockParser   // specify how to parse the block from the api request
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/spec/README.md:107]: done

File: x/spec/README.md [hunk 24]
```diff
@@ -0,0 +1,298 @@
...
+
+```
+type ApiCollection struct {
+	Enabled         bool                // enables/disables the collection
+	CollectionData  CollectionData      // defines the properties of the collection, also acts as a unique key
+	Apis            []*Api              // list of api's in the collection
+	Headers         []*Header           // list of headers supported by the interface and their behaviour
+	InheritanceApis []*CollectionData   // list of other ApiCollection to inherite from
+	ParseDirectives []*[ParseDirective](#parsedirective)   // list of parsing instructions of specific api's
+	Extensions      []*Extension        // list of extensions that providers can support in addition to the basic behaviour (for example, archive node)
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/spec/README.md:70]: done

**Review by oren-lava — APPROVED** [review]

**Pull request merged**